maze_transformer/dataset/maze_dataset.py

```diff
@@ -395,4 +411,173 @@ def cut_percentile_shortest(
         ]
         new_dataset: MazeDataset = MazeDataset(cfg=dataset.cfg, mazes=filtered_mazes)
 
+        return copy.deepcopy(new_dataset)
+
+    @register_dataset_filter
+    @staticmethod
+    def truncate_count(
+        dataset: MazeDataset,
+        max_count: int,
+    ) -> MazeDataset:
+        """truncate the dataset to be at most `max_count` mazes"""
+        new_dataset: MazeDataset = MazeDataset(
+            cfg=dataset.cfg, mazes=dataset.mazes[:max_count]
+        )
+        return copy.deepcopy(new_dataset)
+
+    @register_dataset_filter
+    @staticmethod
+    def remove_duplicates(
+        dataset: MazeDataset,
+        minimum_difference_connection_list: int | None = 1,
+        minimum_difference_solution: int | None = 1,
+    ) -> MazeDataset:
+        """remove duplicates from a dataset, keeping the **LAST** unique maze
+
+        set minimum either minimum difference to `None` to disable checking
+
+        if you want to avoid mazes which have more overlap, set the minimum difference to be greater
+
+        Gotchas:
+        - if two mazes are of different sizes, they will never be considered duplicates
+        - if two solutions are of different lengths, they will never be considered duplicates
+            TODO: check for overlap?
+        """
+        unique_mazes: list[SolvedMaze] = list()
+
+        maze_a: SolvedMaze
+        maze_b: SolvedMaze
+        for i, maze_a in enumerate(dataset.mazes):
+            a_unique: bool = True
+            for maze_b in dataset.mazes[i + 1 :]:
```

Comment:
This is quadratic in the number of mazes, right? I did some quick benchmarking:
 
100 mazes - 40ms
1000 mazes - 3.4s
10000 mazes - 300s

Looks like this will be too slow for our standard 5 million maze dataset. Since we're trying to get this out quickly maybe we should simplify to just consider exact matches for now so we can use a set?

Reply:
Thanks for benchmarking this! I agree this is way too slow. Let's add a `remove_duplicates_fast` filter which does this?

Reply:
Okay I've added this. Dedupes a 100k dataset in 15s and scales ~linearly. But I'm not sure if it is useful. On my 100k dataset there were no exact duplicates for 6x6. I did verify that there are lots of duplicates for 2x2.